src/common/primitive_desc.hpp

```diff
@@ -410,6 +410,7 @@ struct primitive_desc_t : public c_compatible {
         CHECK(_pd->init(engine));
         CHECK(_pd->init_scratchpad_md());
         *pd = _pd.release();
+        DNNL_PRIMITIVE_CREATE(pd_t);
```

Comment:
I'm 100% sure this is the wrong place to fix the problem. At least because this piece doesn't cover all primitives.

Reply:
Any suggestion? I don't find better place to fix this issue.

Reply:
The failure occurred in:
https://github.com/openvinotoolkit/openvino/blob/02f5b7a7e34ef7a41598504d5d6545319bba159b/src/plugins/intel_cpu/src/nodes/deconv.cpp#L1083-L1088
In the failure case, creating` fwd_conv_pd` will use `ref_convolution_fwd_t` but this primitive will not be created and flaged `FACTORY_DNNL,CREATE`, so that in selective build stage `ref_convolution_fwd_t` is not in the kernel_list, and `createDescriptorInternalDefault` failed to create its primitive desc.

Reply:
Since CC is purely branch feature, I have no idea how it's exactly working, but hacking internals disconnected from the feature is definitely a wrong change.
The best thing I was able to find is this:
```
# define DNNL_PRIMITIVE_CREATE(pd_t) OV_ITT_SCOPED_TASK(dnnl::FACTORY_DNNL, std::string("CREATE$CPUEngine$") + typeid(pd_t).name());
```
There's no `OV_ITT_SCOPED_TASK` defined in OneDNN repo so nothing helps to figure it out what's going on under this call.

This sounds like a problem for any nested primitive thing, thus, concat/sum also should fail the same way (since creating reorder) and this change wouldn't cover it.

Nested Primitive is created in host primitive init stage.
I would expect that putting this piece here disrupts WHOLE creation model.
Ultimately, I don't care what happens inside the fork, just trying to keep oneDNN part inside OV a little bit sane.
@dmitry-gorokhov @EgorDuplensky

Reply:
Conditional Compilation(CC) has 2 steps to generate the final package targeting optimized binaries size:

Steps 1: Selective code ITT analyzing to get statistical data

It will enable Macro `SELECTIVE_BUILD_ANALYZER` to build OpenVINO+oneDNN, which leverage a series of CC macros to collect statistical data with the help of ITT profiling:
https://github.com/openvinotoolkit/oneDNN/blob/3767662f257270921b64ec9a40e45a46b4ef048c/src/common/dnnl_sel_build.hpp#L41-L64

Once build is done, run application (for example: benchmark) to runtime collect the statistical data, which will be used to generate a header file named conditional_compilation_gen.h to label which code block need be kept, for example:

```
...
#define ov_pass_PropagatePrecision_run_on_model 1
#define ov_pass_InsertMoveBroadcast 1

#define internal_op_Subgraph 1
#define internal_op_ConvertTruncation_clone_with_new_inputs 1
#define internal_op_LoadConvert_validate_and_infer_types 1

#define intel_cpu_FuseConvolutionAndZeroPoints_ConvNode 1
#define intel_cpu_FuseConvolutionMatMulDeconvAndBias_ParentNode 1
#define intel_cpu_FuseConvolutionMatMulDeconvAndBias_ChildNode 1
#define intel_cpu_FuseDeconvolutionAndSimpleOperation_ParentNode 1
#define intel_cpu_FuseConvolutionAndSimpleOperationThroughMaxPool_ParentNode 1
#define intel_cpu_FuseConvolutionAndSimpleOperation_ParentNode 1
#define intel_cpu_Concat 1
#define intel_cpu_Snippet 1
#define intel_cpu_Convolution 1
#define intel_cpu_Deconvolution 1
#define intel_cpu_Input 1
#define intel_cpu_Eltwise 1
#define intel_cpu_Input 1
#define intel_cpu_Generic 1

#define DNNL_typed_zero_pad_blk_8ba 1
#define DNNL_jit_avx2_dw_convolution_fwd_t 1
#define DNNL_ref_convolution_fwd_t 1
#define DNNL_gemm_convolution_bwd_data_t 1
#define DNNL_x64_jit_blk_reorder_t 1
#define DNNL_jit_avx2_1x1_convolution_fwd_t 1
#define DNNL_jit_avx2_convolution_fwd_t 1
#define DNNL_ref_convolution_bwd_data_t 1
#define DNNL_x64_jit_uni_reorder_t 1
#define DNNL_jit_avx2_fork_dw_convolution_bwd_data_t 1
#define DNNL_gemm_convolution_fwd_t 1
#define DNNL_brdgmm_dw_convolution_fwd_t 1
```

Step 2: Selective build:

It will enable Macro `SELECTIVE_BUILD` and above generated header file to build the final binaries.  The DNNL primitive that not defined in above header files will not be built any more, so the final binaries size become the smallest.

https://github.com/openvinotoolkit/oneDNN/blob/3767662f257270921b64ec9a40e45a46b4ef048c/src/common/dnnl_sel_build.hpp#L65-L86


Above is CC working flow.

we have added `DNNL_PRIMITIVE_CREATE` during primitive creating to label it should be kept:
https://github.com/openvinotoolkit/oneDNN/blob/3767662f257270921b64ec9a40e45a46b4ef048c/src/common/primitive_desc.hpp#L428-L434

But seems it has missed a case, sometimes only need this primitive's desc but not need create this primitive in final execution graph. In this case we need also label this primitive. It means that if one primitive desc has ever been created successfully during CC analyzing step, it should be kept to let CC step 2 package also can create its primitive desc. So this PR add a label if the primitive desc is created successfully.


About Nested Primitive issue, I don't think it is a problem, because `DNNL_PRIMITIVE_CREATE` only take effect during CC analyzing stage and add a label into the statistic data, not impact primitive and whole model. And it will be set empty at the final selective build stage.


For concat/sum primitive, we also can add such label Macro at the same pd init() method, but it seems that openvino cpu plugin doesn't create their primitive desc if not need. But for deconv primitive it need to create its fwd_conv_pd:
https://github.com/openvinotoolkit/openvino/blob/02f5b7a7e34ef7a41598504d5d6545319bba159b/src/plugins/intel_cpu/src/nodes/deconv.cpp#L1083-L1088

I'm not sure we need do similar thing for all pd init() method @dmatveev @ilya-lavrenov any any comments?

Reply:
Do you still plan to stick to this change?

Given that CC is likely looking for a impl name, any issues with parsing a second part of the top-level impl name and include all relevant convolutions (in case of deconvolution), and same for other nested primitives?